### PR TITLE
Text support

### PR DESCRIPTION
**Pull request by prosif**
Add text support + support for floating point size/position values. Increases entity frame size from 8 indices to 46- original 8 plus 4 (+1 for two values after decimal for each size/position value), then 34 for text. 2 for x, y, then 32 for ASCII values of text. This can/will definitely be optimized, as not every frame needs to be so large since not every entity has text. But basic text support is here.

### CODE REVIEW

File: src/Squisher.js [hunk 1]
```diff
@@ -52,16 +52,29 @@ class Squisher {
...
+            this.pixelBoard[entityFrameSize * i + 2] = this.entities[i].color[2];
+            this.pixelBoard[entityFrameSize * i + 3] = this.entities[i].color[3];
+            this.pixelBoard[entityFrameSize * i + 4] = Math.floor(this.entities[i].pos.x * 100);
+            this.pixelBoard[entityFrameSize * i + 5] = this.entities[i].pos.x * 10000 % 100;
+            this.pixelBoard[entityFrameSize * i + 6] = Math.floor(this.entities[i].pos.y * 100);
+            this.pixelBoard[entityFrameSize * i + 7] = this.entities[i].pos.y * 10000 % 100;
+            this.pixelBoard[entityFrameSize * i + 8] = Math.floor(this.entities[i].size.x * 100);
+            this.pixelBoard[entityFrameSize * i + 9] = this.entities[i].size.x * 10000 % 100;
+            this.pixelBoard[entityFrameSize * i + 10] = Math.floor(this.entities[i].size.y * 100);
+            this.pixelBoard[entityFrameSize * i + 11] = this.entities[i].size.y * 10000 % 100;
```
Comment (YazeedLoonat) [inline src/Squisher.js]: I'm not 1 hundo on this, 
looking at: 
https://www.khanacademy.org/computing/computer-science/cryptography/modarithmetic/a/modular-multiplication 
we get:
A * 10000 % 100 =  (A % 100 * 10000 % 100) % 100 =  (A % 100 * 0) % 100 = 0
Where A is any number

**Comment by prosif** [comment]
I put up this PR as a rough v1 for myself, I didn't intend for another person to actually read it. But still, thanks for taking a look! 

I still need to add some stuff to this branch, including:

- text size
- text font
- optimization for entity frame size when they have no text

**Comment by prosif** [comment]
I've updated this PR with a whole lot more stuff that includes support for SSL, audio, text & images.

This is still very much a preliminary PR, I just want to have all of the changes in one big thing with comments/changes over time. I will work on cleaning this PR up and fixing bugs, then moving on to the new data structure to store entity state in the squisher. 

Until then, this is pretty rad.

**Comment by prosif** [comment]
And now I've read my (very rude sounding) previous comment, which reminds me that I need to add config for text as well. Sorry about that comment btw

**Comment by prosif** [comment]
Alright so now I've made a couple more big changes. 

1. Removed SSL, as I realized I didn't need it anymore. Or at least I don't think so. Anyway it's a relief because I don't know how we'd set up SSL for every homegames instance. 
2. Replaced usage of `192..` with local hostname acquired from URL box
3. Added support for `homegames.link` to allow easy mapping to the local IP without having to physically type out your IP address (gross)
4. A bit of general cleanup with server stuff.

I still need to clean up what's here before it's ready to review & merge.

**Comment by prosif** [comment]
Also, check out [homegames.link](https://github.com/prosif/homegames.link) to see what that is. You should be able to use it locally when testing @YazeedLoonat, let me know if it doesn't work. 

I've also updated http://picodeg.io to use this branch

File: src/Player.js [hunk 2]
```diff
@@ -1,29 +1,35 @@
...
         }
         
-        for (let listener of this.inputListeners) {
+        for (const listener of this.inputListeners) {
             listener.handlePlayerInput(this, data);
         }
     }
 
     disconnect() {
-        // handle cleanup here. will probably need to notify listeners. too many listeners.
```
Comment (prosif) [inline src/Player.js]: Too true, past me. Too true.

File: src/Squisher.js [hunk 3]
```diff
@@ -93,13 +122,80 @@ class Squisher {
         }
         const entity = this.clickListeners[click.x * this.width + click.y];
         if (entity) {
-            entity.handleClick(player, translatedX, translatedY);
+            entity.handleClick && entity.handleClick(player, translatedX, translatedY);
         }
     }
+    
+    squish(entity) {
+        // Type (1) + Size (1) + color (4) + pos (4) + size (4) + text position (2) + text (32) + assets (37 * assetCount)
```
Comment (prosif) [inline src/Squisher.js]: This function is gross and I don't like it. I want to clean this up but I'm still making big changes to what's sent over the socket so I will wait until that's a little more clear before abstracting this logic

File: src/games/draw.js [hunk 4]
```diff
@@ -1,17 +1,52 @@
...
+const gameNode = require("../GameNode");
+const {colors, randomColor} = require("../Colors");
 const colorKeys = Object.keys(colors);
+const Asset = require("../Asset");
 
 class Draw {
     constructor() {
+        this.assets = {
+            "test": new Asset("url", {
+                "location": "https://www.nicepng.com/png/full/323-3239506_kanye-west-shrug-transparent.png"
```
Comment (prosif) [inline src/games/draw.js]: nice

File: src/Squisher.js [hunk 5]
```diff
@@ -17,7 +19,39 @@ class Squisher {
...
+            while (keyIndex < assetKeyLength && keyIndex < key.length) {
+                assetKeyArray[keyIndex] = key.charCodeAt(keyIndex);
+                keyIndex++;
+            }
+
+            const encodedLength = (payload.length + assetKeyLength).toString(36);
+            
+            const assetType = gameAssets[key].info.type === "image" ? 1 : 2;
+
+            this.assets[key] = [ASSET_TYPE, assetType, encodedLength.charCodeAt(0), encodedLength.charCodeAt(1), encodedLength.charCodeAt(2), encodedLength.charCodeAt(3), ...assetKeyArray, ...payload];
```
Comment (prosif) [inline src/Squisher.js]: This logic should be abstracted. Pretty sure we won't support video in the near future (though I suppose that's not too crazy), so image/audio are probably enough for assets for now

File: src/games/draw.js [hunk 6]
```diff
@@ -1,17 +1,49 @@
...
+const gameNode = require("../GameNode");
+const {colors, randomColor} = require("../Colors");
 const colorKeys = Object.keys(colors);
+const Asset = require("../Asset");
 
 class Draw {
     constructor() {
+        this.assets = {
+            "test": new Asset("url", {
+                "location": "https://www.nicepng.com/png/full/323-3239506_kanye-west-shrug-transparent.png",
```
Comment (prosif) [inline src/games/draw.js]: nice

File: web/App.js [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+socket.onmessage = function(msg) {
+    const buf = new Uint8ClampedArray(msg.data);
+    let color, startX, startY, width, height;
+    let i = 0;
+    while (i < buf.length) {
+        const frameType = buf[i];
+
+        if (frameType === 1) {
+            const assetType = buf[i + 1];
```
Comment (prosif) [inline web/App.js]: This is a weird combination of the inverse of the "squish" function in `Squisher.js` & asset decoding/storing. 

I would like to have one common module that can be shared by both the web app & the server that will squish/unsquish classes of thing. Kind of like a common "game node serializer". That's the intent of Squisher, really.

File: web/App.js [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+}
+
+document.addEventListener('touchstart', unlock, false);
+
+canvas.addEventListener('mousedown', function(e) {
+    mouseDown = true;
+    unlock();
+});
+
+canvas.addEventListener('mouseup', function(e) {
```
Comment (prosif) [inline web/App.js]: Once this list of event listeners grows and starts to split up more based on device, we should abstract some of this stuff

File: web/App.js [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+    e.preventDefault();
+    click(e.touches['0'].clientX, e.touches['0'].clientY);
+});
+
+function keyMatters(event) {
+    // Key code values 36-40 are the arrow keys
+    return event.key.length == 1 && event.key >= ' ' && event.key <= 'z' || event.keyCode >= 36 && event.keyCode <= 40;
+}
+
+document.addEventListener('keydown', function(e) {
```
Comment (prosif) [inline web/App.js]: TODO: Handle CMD+R or F5 to refresh.

**Pull request merged**